File: android/app/src/main/java/com/localllmapp/DeviceControlModule.java
package com.localllmapp;

import android.app.Activity;
import android.content.Context;
import android.content.ClipboardManager;
import android.content.ClipData;
import android.content.Intent;
import android.net.Uri;
import android.database.Cursor;
import android.provider.OpenableColumns;
import android.speech.RecognizerIntent;
import java.io.InputStream;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;

import com.facebook.react.bridge.ActivityEventListener;
import com.facebook.react.bridge.BaseActivityEventListener;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.Arguments;

public class DeviceControlModule extends ReactContextBaseJavaModule {

    private static final int SPEECH_REQUEST_CODE = 100;
    private static final int FILE_REQUEST_CODE = 101;
    private static final long MAX_TEXT_FILE_BYTES = 512 * 1024;
    private Promise speechPromise;
    private Promise filePromise;

    private final ActivityEventListener activityEventListener = new BaseActivityEventListener() {
        @Override
        public void onActivityResult(Activity activity, int requestCode, int resultCode, Intent data) {
            if (requestCode == SPEECH_REQUEST_CODE && speechPromise != null) {
                if (resultCode == Activity.RESULT_OK && data != null) {
                    ArrayList<String> results = data.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS);
                    if (results != null && !results.isEmpty()) {
                        speechPromise.resolve(results.get(0));
                    } else {
                        speechPromise.reject("NO_RESULT", "No speech recognized");
                    }
                } else {
                    speechPromise.reject("CANCELLED", "Speech recognition cancelled");
                }
                speechPromise = null;
            } else if (requestCode == FILE_REQUEST_CODE && filePromise != null) {
                if (resultCode == Activity.RESULT_OK && data != null) {
                    Uri uri = data.getData();
                    if (uri != null) {
                        try {
                            String displayName = "unknown.txt";
                            long sizeBytes = 0;
                            try (Cursor cursor = activity.getContentResolver().query(uri, null, null, null, null)) {
                                if (cursor != null && cursor.moveToFirst()) {
                                    int nameIndex = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME);
                                    int sizeIndex = cursor.getColumnIndex(OpenableColumns.SIZE);
                                    if (nameIndex != -1) {
                                        String queriedName = cursor.getString(nameIndex);
                                        if (queriedName != null && !queriedName.trim().isEmpty()) {
                                            displayName = queriedName;
                                        }
                                    }
                                    if (sizeIndex != -1 && !cursor.isNull(sizeIndex)) {
                                        sizeBytes = cursor.getLong(sizeIndex);
                                    }
                                }
                            }

                            String sizeStr = sizeBytes > 1024 ? (sizeBytes / 1024) + " KB" : sizeBytes + " B";
                            String lowerName = displayName.toLowerCase();
                            String contentStr;

                            if (lowerName.endsWith(".jpg") || lowerName.endsWith(".jpeg") || lowerName.endsWith(".png") || lowerName.endsWith(".gif") || lowerName.endsWith(".webp") || lowerName.endsWith(".bmp") || lowerName.endsWith(".pdf") || lowerName.endsWith(".zip")) {
                                contentStr = "[Attached File: " + displayName + " (" + sizeStr + ")]\nNote: This is an image or binary document. Local text-only GGUF models cannot visually inspect image pixels directly. If there is text in this image/document you would like summarized or analyzed, please copy or type the relevant text directly into the chat.";
                            } else if (sizeBytes > MAX_TEXT_FILE_BYTES) {
                                contentStr = "[Attached File: " + displayName + " (" + sizeStr + ")]\nThis text file is larger than the 512 KB attachment limit. Please paste a smaller excerpt into chat.";
                            } else {
                                StringBuilder sb = new StringBuilder();
                                try (InputStream is = activity.getContentResolver().openInputStream(uri)) {
                                    if (is == null) {
                                        throw new IllegalStateException("Unable to open selected file");
                                    }
                                    BufferedReader reader = new BufferedReader(new InputStreamReader(is, "UTF-8"));
                                    String line;
                                    int lineCount = 0;
                                    int charCount = 0;
                                    while ((line = reader.readLine()) != null && lineCount < 2000 && charCount < MAX_TEXT_FILE_BYTES) {
                                        String cleanLine = line.replaceAll("[\\u0000-\\u0008\\u000B\\u000C\\u000E-\\u001F]", "");
                                        sb.append(cleanLine).append("\n");
                                        charCount += cleanLine.length() + 1;
                                        lineCount++;
                                    }
                                }
                                contentStr = sb.toString();
                            }

                            WritableMap map = Arguments.createMap();
                            map.putString("name", displayName);
                            map.putString("size", sizeStr);
                            map.putString("content", contentStr);
                            map.putString("uri", uri.toString());
                            filePromise.resolve(map);
                        } catch (Exception e) {
                            filePromise.reject("READ_ERROR", "Failed to read selected file: " + e.getMessage());
                        }
                    } else {
                        filePromise.reject("NO_URI", "No file selected");
                    }
                } else {
                    filePromise.reject("CANCELLED", "File selection cancelled");
                }
                filePromise = null;
            }
        }
    };

    public DeviceControlModule(ReactApplicationContext reactContext) {
        super(reactContext);
        reactContext.addActivityEventListener(activityEventListener);
    }

    @Override
    public String getName() {
        return "DeviceControl";
    }

    @ReactMethod
    public void startSpeechRecognition(Promise promise) {
        Activity currentActivity = getCurrentActivity();
        if (currentActivity == null) {
            promise.reject("NO_ACTIVITY", "No activity available");
            return;
        }
        if (speechPromise != null) {
            promise.reject("REQUEST_IN_PROGRESS", "Speech recognition is already active");
            return;
        }
        this.speechPromise = promise;
        Intent intent = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
        intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM);
        intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE, "en-US");
        intent.putExtra(RecognizerIntent.EXTRA_PROMPT, "Speak to Moonlight...");
        intent.putExtra(RecognizerIntent.EXTRA_MAX_RESULTS, 1);
        try {
            currentActivity.startActivityForResult(intent, SPEECH_REQUEST_CODE);
        } catch (Exception e) {
            this.speechPromise = null;
            promise.reject("SPEECH_ERROR", "Could not launch speech recognition: " + e.getMessage());
        }
    }

    @ReactMethod
    public void pickFile(Promise promise) {
        Activity currentActivity = getCurrentActivity();
        if (currentActivity == null) {
            promise.reject("NO_ACTIVITY", "No activity available");
            return;
        }
        if (filePromise != null) {
            promise.reject("REQUEST_IN_PROGRESS", "File picker is already active");
            return;
        }
        this.filePromise = promise;
        Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
        intent.setType("*/*");
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        try {
            currentActivity.startActivityForResult(Intent.createChooser(intent, "Select File for Moonlight AI"), FILE_REQUEST_CODE);
        } catch (Exception e) {
            promise.reject("PICKER_ERROR", "Could not launch file picker: " + e.getMessage());
            this.filePromise = null;
        }
    }

    @ReactMethod
    public void copyToClipboard(String text, Promise promise) {
        try {
            ClipboardManager clipboard = (ClipboardManager) getReactApplicationContext().getSystemService(Context.CLIPBOARD_SERVICE);
            ClipData clip = ClipData.newPlainText("Moonlight AI", text != null ? text : "");
            if (clipboard != null) {
                clipboard.setPrimaryClip(clip);
            }
            promise.resolve(true);
        } catch (Exception e) {
            promise.reject("CLIPBOARD_ERROR", "Failed to copy: " + e.getMessage());
        }
    }
}

File: android/app/src/main/java/com/moonknightstudio/moonlightai/DeviceControlModule.kt
package com.moonknightstudio.moonlightai

import android.app.Activity
import android.app.ActivityManager
import android.os.StatFs
import android.os.Build
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.database.Cursor
import android.net.Uri
import android.provider.OpenableColumns
import android.speech.RecognizerIntent
import com.facebook.react.bridge.ActivityEventListener
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.BaseActivityEventListener
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import java.util.Locale
import java.net.URL
import javax.net.ssl.HttpsURLConnection
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import java.io.ByteArrayOutputStream
import org.json.JSONObject

class DeviceControlModule(
  private val appContext: ReactApplicationContext,
) : ReactContextBaseJavaModule(appContext) {
  companion object {
    private const val SPEECH_REQUEST_CODE = 4100
    private const val FILE_REQUEST_CODE = 4101
    private val MAX_TEXT_FILE_BYTES = BoundedTextReader.MAX_BYTES.toLong()
  }

  private var speechPromise: Promise? = null
  private val searchCredentials = SearchCredentialStore(appContext)
  private val credentialExecutor = java.util.concurrent.Executors.newSingleThreadExecutor()

  @ReactMethod
  fun saveSearchCredentials(value: String, promise: Promise) {
    credentialExecutor.execute {
      try { searchCredentials.save(value); promise.resolve(true) }
      catch (_: Exception) { promise.reject("SEARCH_STORAGE", "Could not securely save search access. Try again.") }
    }
  }
  @ReactMethod
  fun readSearchCredentials(promise: Promise) {
    credentialExecutor.execute {
      try { promise.resolve(searchCredentials.read()) }
      catch (_: Exception) { promise.reject("SEARCH_STORAGE", "Saved search access could not be unlocked. Re-enter your key in Settings.") }
    }
  }
  @ReactMethod
  fun clearSearchCredentials(promise: Promise) {
    credentialExecutor.execute {
      try { searchCredentials.clear(); promise.resolve(true) }
      catch (_: Exception) { promise.reject("SEARCH_STORAGE", "Could not remove saved search access. Try again.") }
    }
  }
  private var filePromise: Promise? = null
  private data class SearchRequest(val connection: HttpsURLConnection, val cancelled: AtomicBoolean = AtomicBoolean(false))
  private val searches = ConcurrentHashMap<String, SearchRequest>()

  @ReactMethod
  fun cancelWebSearch(requestId: String) {
    searches[requestId]?.let { it.cancelled.set(true); it.connection.disconnect() }
  }

  @ReactMethod
  fun requestWebSearch(requestId: String, endpoint: String, code: String, query: String, promise: Promise) {
    try {
      val url = URL(endpoint)
      require(url.protocol == "https" && url.userInfo == null && url.query == null && url.ref == null)
      require(query.isNotBlank() && query.length <= 400 && code.matches(Regex("[A-Za-z0-9_-]{32,128}")))
      require(searches.isEmpty()) { "A search is already running" }
      val connection = url.openConnection() as HttpsURLConnection
      val request = SearchRequest(connection)
      searches[requestId] = request
      Thread {
        try {
          if (request.cancelled.get()) throw IllegalStateException("Cancelled")
          connection.instanceFollowRedirects = false
          connection.connectTimeout = 10000
          connection.readTimeout = 10000
          connection.requestMethod = "POST"
          connection.doOutput = true
          connection.setRequestProperty("Content-Type", "application/json")
          connection.setRequestProperty("Authorization", "Bearer $code")
          val payload = JSONObject().put("query", query).toString().toByteArray(Charsets.UTF_8)
          connection.setFixedLengthStreamingMode(payload.size)
          connection.outputStream.use { it.write(payload) }
          val status = connection.responseCode
          val output = ByteArrayOutputStream()
          // Never follow redirects with tester credentials, or buffer unbounded responses.
          if (status == 200) connection.inputStream.use { stream ->
            val buffer = ByteArray(2048)
            while (true) {
              if (request.cancelled.get()) throw IllegalStateException("Cancelled")
              val count = stream.read(buffer)
              if (count < 0) break
              require(output.size() + count <= 16000) { "Response too large" }
              output.write(buffer, 0, count)
            }
          }
          if (request.cancelled.get()) throw IllegalStateException("Cancelled")
          val result = Arguments.createMap()
          result.putInt("status", status)
          result.putString("body", output.toString("UTF-8"))
          promise.resolve(result)
        } catch (error: Exception) {
          promise.reject("SEARCH_FAILED", "Search could not complete. Check the connection or try again.")
        } finally {
          connection.disconnect()
          searches.remove(requestId)
        }
      }.start()
    } catch (error: Exception) { promise.reject("SEARCH_CONFIG", "Search connection is invalid or busy.") }
  }

  private val activityEventListener: ActivityEventListener =
    object : BaseActivityEventListener() {
      override fun onActivityResult(
        activity: Activity,
        requestCode: Int,
        resultCode: Int,
        data: Intent?,
      ) {
        when (requestCode) {
          SPEECH_REQUEST_CODE -> finishSpeechRequest(resultCode, data)
          FILE_REQUEST_CODE -> finishFileRequest(activity, resultCode, data)
        }
      }
    }

  init {
    appContext.addActivityEventListener(activityEventListener)
  }

  override fun getName(): String = "DeviceControl"

  @ReactMethod
  fun getDeviceCapacity(promise: Promise) {
    try {
      val manager = appContext.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
      val info = ActivityManager.MemoryInfo()
      manager.getMemoryInfo(info)
      val disk = StatFs(appContext.filesDir.absolutePath)
      promise.resolve(Arguments.createMap().apply {
        putDouble("totalMemory", info.totalMem.toDouble())
        putDouble("availableMemory", info.availMem.toDouble())
        putDouble("freeStorage", disk.availableBytes.toDouble())
        putBoolean("is64Bit", Build.SUPPORTED_64_BIT_ABIS.isNotEmpty())
      })
    } catch (error: Exception) {
      promise.reject("CAPACITY_ERROR", "Could not check device capacity.", error)
    }
  }

  @ReactMethod
  fun startSpeechRecognition(promise: Promise) {
    val activity = appContext.currentActivity
    if (activity == null) {
      promise.reject("NO_ACTIVITY", "Voice input is unavailable because the app is not active.")
      return
    }
    if (speechPromise != null) {
      promise.reject("REQUEST_IN_PROGRESS", "Voice input is already active.")
      return
    }

    val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).apply {
      putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM)
      putExtra(RecognizerIntent.EXTRA_LANGUAGE, Locale.getDefault().toLanguageTag())
      putExtra(RecognizerIntent.EXTRA_PROMPT, "Speak to Moonlight AI")
      putExtra(RecognizerIntent.EXTRA_MAX_RESULTS, 1)
    }
    if (intent.resolveActivity(activity.packageManager) == null) {
      promise.reject(
        "SPEECH_UNAVAILABLE",
        "No speech-recognition service is available on this device.",
      )
      return
    }

    speechPromise = promise
    try {
      activity.startActivityForResult(intent, SPEECH_REQUEST_CODE)
    } catch (error: Exception) {
      speechPromise = null
      promise.reject("SPEECH_ERROR", "Could not start voice input.", error)
    }
  }

  private fun finishSpeechRequest(resultCode: Int, data: Intent?) {
    val promise = speechPromise ?: return
    speechPromise = null
    if (resultCode != Activity.RESULT_OK || data == null) {
      promise.reject("CANCELLED", "Voice input was cancelled.")
      return
    }

    val result = data.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS)?.firstOrNull()
    if (result.isNullOrBlank()) {
      promise.reject("NO_RESULT", "No speech was recognized. Please try again.")
    } else {
      promise.resolve(result)
    }
  }

  @ReactMethod
  fun pickFile(promise: Promise) {
    val activity = appContext.currentActivity
    if (activity == null) {
      promise.reject("NO_ACTIVITY", "The file picker is unavailable because the app is not active.")
      return
    }
    if (filePromise != null) {
      promise.reject("REQUEST_IN_PROGRESS", "The file picker is already active.")
      return
    }

    val intent = Intent(Intent.ACTION_GET_CONTENT).apply {
      type = "*/*"
      addCategory(Intent.CATEGORY_OPENABLE)
    }
    if (intent.resolveActivity(activity.packageManager) == null) {
      promise.reject("PICKER_UNAVAILABLE", "No compatible file picker is installed.")
      return
    }

    filePromise = promise
    try {
      activity.startActivityForResult(
        Intent.createChooser(intent, "Select a file for Moonlight AI"),
        FILE_REQUEST_CODE,
      )
    } catch (error: Exception) {
      filePromise = null
      promise.reject("PICKER_ERROR", "Could not open the file picker.", error)
    }
  }

  private fun finishFileRequest(activity: Activity, resultCode: Int, data: Intent?) {
    val promise = filePromise ?: return
    filePromise = null
    if (resultCode != Activity.RESULT_OK || data?.data == null) {
      promise.reject("CANCELLED", "File selection was cancelled.")
      return
    }

    try {
      promise.resolve(readSelectedFile(activity, data.data!!))
    } catch (error: Exception) {
      promise.reject("READ_ERROR", "Failed to read the selected file.", error)
    }
  }

  private fun readSelectedFile(activity: Activity, uri: Uri) =
    Arguments.createMap().apply {
      var displayName = "selected-file"
      var sizeBytes = 0L
      activity.contentResolver.query(uri, null, null, null, null)?.use { cursor: Cursor ->
        if (cursor.moveToFirst()) {
          cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME).takeIf { it >= 0 }?.let {
            displayName = cursor.getString(it)?.takeIf(String::isNotBlank) ?: displayName
          }
          cursor.getColumnIndex(OpenableColumns.SIZE).takeIf { it >= 0 && !cursor.isNull(it) }?.let {
            sizeBytes = cursor.getLong(it)
          }
        }
      }

      val sizeLabel = if (sizeBytes > 1024) "${sizeBytes / 1024} KB" else "$sizeBytes B"
      val lowerName = displayName.lowercase(Locale.ROOT)
      val binaryExtensions = listOf(".jpg", ".jpeg", ".png", ".gif", ".webp", ".bmp", ".pdf", ".zip")
      val content = when {
        binaryExtensions.any(lowerName::endsWith) ->
          "[Attached File: $displayName ($sizeLabel)]\nThis binary file cannot be read by the local text-only model. Paste any relevant text into the chat."
        sizeBytes > MAX_TEXT_FILE_BYTES ->
          "[Attached File: $displayName ($sizeLabel)]\nThis file is larger than the 512 KB text attachment limit. Paste a smaller excerpt into the chat."
        else -> readTextContent(activity, uri)
      }

      putString("name", displayName)
      putString("size", sizeLabel)
      putString("content", content)
      putString("uri", uri.toString())
    }

  private fun readTextContent(activity: Activity, uri: Uri): String {
    val input = activity.contentResolver.openInputStream(uri)
      ?: throw IllegalStateException("Unable to open the selected file")
    return input.use { stream -> BoundedTextReader.read(stream) }
  }

  @ReactMethod
  fun copyToClipboard(text: String?, promise: Promise) {
    try {
      val clipboard = appContext.getSystemService(Context.CLIPBOARD_SERVICE) as? ClipboardManager
      if (clipboard == null) {
        promise.reject("CLIPBOARD_UNAVAILABLE", "The clipboard service is unavailable.")
        return
      }
      clipboard.setPrimaryClip(ClipData.newPlainText("Moonlight AI", text.orEmpty()))
      promise.resolve(true)
    } catch (error: Exception) {
      promise.reject("CLIPBOARD_ERROR", "Failed to copy text.", error)
    }
  }

  override fun invalidate() {
    appContext.removeActivityEventListener(activityEventListener)
    speechPromise?.reject("MODULE_DESTROYED", "Voice input stopped because the app closed.")
    filePromise?.reject("MODULE_DESTROYED", "File selection stopped because the app closed.")
    speechPromise = null
    filePromise = null
    super.invalidate()
  }
}
